web/client/epics/layerfilter.js

```diff
@@ -114,8 +118,139 @@ export const onApplyFilter = (action$, {getState}) =>
+ */
+export const applyWMSLegendFilterForMap = (action$, {getState}) => {
+    return  action$.ofType(LAYER_FILTER_BY_LEGEND).
+        filter(({layerId}) => {
+            const state = getState();
+            const layer = head(state?.layers?.flat.filter(l => l.id === layerId));
+            const isMap = mapIdSelector(state);
+
+            return layer?.type  === 'wms' && isMap;
+        })
+        .switchMap((action) => {
+            const state = getState();
+            const layer = head(state.layers.flat.filter(l => l.id === action.layerId));
+            const queryFormFilterObj = {...get(getState(), "queryform", {})};
+            let filterObj = layer.layerFilter ? layer.layerFilter : queryFormFilterObj;
+            let searchUrl = layer?.search?.url;
+            // reset thte filter if legendCQLFilter is empty
+            if (!action.legendCQLFilter) {
+                // clear legend filter with id = 'interactiveLegend'
+                const isLegendFilterExist = filterObj?.filters?.find(f => f.id === 'interactiveLegend');
+                if (isLegendFilterExist) {
+                    filterObj = {
+                        ...filterObj, filters: filterObj?.filters?.filter(f => f.id !== 'interactiveLegend')
+                    };
+                }
+                let newFilter = isNotEmptyFilter(filterObj) ? filterObj : undefined;
+                return Rx.Observable.of(search(searchUrl, newFilter), addFilterToLayer(layer.id, newFilter), storeAppliedFilter(newFilter));
+            }
+            let cqlStatement = action.legendCQLFilter;
+            if (cqlStatement) {
+                // remove the unnecessaru brackets --> in some cases wms getLegendGraphic provides filter with brackets
+                // "[FIELD_01 >= '1' AND FIELD_01 < '0']" ---> to be "FIELD_01 >= '1' AND FIELD_01 < '0'"
+                cqlStatement = cqlStatement.includes("[") ? cqlStatement.slice(1, cqlStatement.length - 1) : cqlStatement;
+            }
+            // add legend filter with id = 'interactiveLegend' in cql format
+            let filter = {
+                ...(filterObj || {}), filters: [
+                    ...(filterObj?.filters?.filter(f => f.id !== 'interactiveLegend') || []), ...[
+                        {
+                            "id": "interactiveLegend",
+                            "format": "logic",
+                            "version": "1.0.0",
+                            "logic": "AND",
+                            "filters": [{
+                                "format": "cql",
+                                "version": "1.0.0",
+                                "body": cqlStatement,
+                                "id": `[${cqlStatement}]`
+                            }]
+                        }
+                    ]
+                ]
+            };
+            return Rx.Observable.of(search(searchUrl, filter), addFilterToLayer(layer.id, filter), storeAppliedFilter(filter));
```

Comment:
After we estracted the filter logic we can remove this epic because the main action we needed addFilterToLayer, it's already covered by the onChange action